### Victoria Ten
#### MMGMT 722, Data Analytics using Python
#### Assignment 1: Finding the Best Delivery Promise for Rosa's Pizza

GitHub Repository Link: [https://github.com/vicarten/rosa-pizza-app](https://github.com/vicarten/rosa-pizza-app)

Streamlit Link: [https://tenv-rosa-pizza-app.streamlit.app/](https://tenv-rosa-pizza-app.streamlit.app/)

## Importing Data

In [13]:
!pip install -q git+https://github.com/zhouy185/rosa-starter.git

In [14]:
import numpy as np
from starter import ZONES, TIME_BLOCKS, COSTS, PROMISE, delivery_times

In [15]:
print(ZONES)
print(TIME_BLOCKS)
print(COSTS)
print(PROMISE)

['Central', 'North', 'Far West']
['Lunch', 'Weekday eve', 'Fri/Sat eve', 'Other']
{'refund': 10.0, 'churn_orders': 1.8, 'margin': 9.0}
45


In [16]:
times = delivery_times("Far West", "Fri/Sat eve", 45)
print(len(times), "orders; average delivery time", round(times.mean(), 1),
"minutes")

210 orders; average delivery time 46.6 minutes


### I(a)
This function returns the percentage of orders delivered later than the promised time. If the zone or time block is 'all', it loops over every zone or time block. Otherwise it uses a one-item list.

In [17]:
def late_percentage(zone, time_block, promise):
  total = 0
  late = 0
  if zone == "all":
    zone = ZONES
  else:
    zone = [zone]
  if time_block == "all":
    time_block = TIME_BLOCKS
  else:
    time_block = [time_block]
  for z in zone:
    for tb in time_block:
      orders = delivery_times(z, tb, promise, seed = 1)
      for o in orders:
        total += 1
        if o > promise:
         late += 1
  percentage = late/total*100
  return round(percentage,2)

print(late_percentage("Far West", "Fri/Sat eve", 45))
print(late_percentage("all", "Lunch", 45))
print(late_percentage("North", "all", 45))
print(late_percentage("all", "all", 45))

36.67
3.33
6.8
8.29


### I(b)
This code calculates the late rate for all 12 zone and time block pairs, stores each result in a list, and sorts the list from highest to lowest late rate.

In [18]:
late_orders = []
for zone in ZONES:
  for time_block in TIME_BLOCKS:
    total = 0
    late = 0
    promise = 45
    orders = delivery_times(zone, time_block, promise, seed = 1)
    for order in orders:
      total += 1
      if order > promise:
       late += 1
    percentage = round(late/total*100, 2)
    late_orders.append((zone, time_block, percentage))
late_orders_sorted = sorted(late_orders, key=lambda item: item[2], reverse=True)
for z, tb, lp in late_orders_sorted:
    print(f"Zone: {z}\tTime Block: {tb}\tLate Orders: {lp}%")

Zone: Far West	Time Block: Fri/Sat eve	Late Orders: 36.67%
Zone: North	Time Block: Fri/Sat eve	Late Orders: 16.67%
Zone: Far West	Time Block: Weekday eve	Late Orders: 11.18%
Zone: Central	Time Block: Fri/Sat eve	Late Orders: 6.84%
Zone: North	Time Block: Weekday eve	Late Orders: 6.0%
Zone: Far West	Time Block: Other	Late Orders: 4.69%
Zone: Central	Time Block: Weekday eve	Late Orders: 4.24%
Zone: North	Time Block: Lunch	Late Orders: 3.68%
Zone: Central	Time Block: Lunch	Late Orders: 3.46%
Zone: Central	Time Block: Other	Late Orders: 2.63%
Zone: North	Time Block: Other	Late Orders: 2.5%
Zone: Far West	Time Block: Lunch	Late Orders: 2.5%


### I(c)
This function returns the average delivery time. It adds up the delivery time of every order in the selected zone and time block, then divides by the number of orders. If the zone or time block is 'all', it includes all of them.

In [19]:
def avg_time(zone, time_block, promise):
  total = 0
  time = 0
  if zone == "all":
    zone = ZONES
  else:
    zone = [zone]
  if time_block == "all":
    time_block = TIME_BLOCKS
  else:
    time_block = [time_block]
  for z in zone:
    for tb in time_block:
      orders = delivery_times(z, tb, promise, seed = 1)
      for o in orders:
        total += 1
        time += o
  avg = time/total
  return round(avg, 2)

print(avg_time("Far West", "Fri/Sat eve", 45))
print(avg_time("all", "Lunch", 45))
print(avg_time("North", "all", 45))
print(avg_time("all", "all", 45))

42.92
28.22
30.98
31.19


### I(d)
This code calculates the average delivery time for all 12 zone and time block pairs, stores each result in a list, and sorts the list from highest to lowest average.

In [20]:
delivery_avg_time = []
for zone in ZONES:
  for time_block in TIME_BLOCKS:
    total = 0
    time = 0
    promise = 45
    orders = delivery_times(zone, time_block, promise, seed = 1)
    for order in orders:
      total += 1
      time += order
    avg = round(time/total, 2)
    delivery_avg_time.append((zone, time_block, avg))
delivery_avg_time_sorted = sorted(delivery_avg_time, key=lambda item: item[2], reverse=True)
for z, tb, lp in delivery_avg_time_sorted:
    print(f"Zone: {z}\tTime Block: {tb}\tAverage delivery time: {lp}minutes")

Zone: Far West	Time Block: Fri/Sat eve	Average delivery time: 42.92minutes
Zone: North	Time Block: Fri/Sat eve	Average delivery time: 36.77minutes
Zone: Far West	Time Block: Weekday eve	Average delivery time: 35.04minutes
Zone: Central	Time Block: Fri/Sat eve	Average delivery time: 32.13minutes
Zone: North	Time Block: Weekday eve	Average delivery time: 31.38minutes
Zone: Far West	Time Block: Lunch	Average delivery time: 30.58minutes
Zone: Far West	Time Block: Other	Average delivery time: 30.04minutes
Zone: Central	Time Block: Weekday eve	Average delivery time: 29.03minutes
Zone: North	Time Block: Lunch	Average delivery time: 28.67minutes
Zone: North	Time Block: Other	Average delivery time: 27.66minutes
Zone: Central	Time Block: Lunch	Average delivery time: 26.8minutes
Zone: Central	Time Block: Other	Average delivery time: 26.04minutes


### I(e)
The late rate is more relevant than the average delivery time for Rosa's decisions. The average delivery time didn't show any problems, since the average is below the promised 45 minutes in every category. The late rate, however, shows that categories differ a lot in how many orders arrive late, and that Far West on Friday and Saturday evenings is the most problematic. For example, its average is 42.92min (which is under promised 45min), yet 36.67% of its orders are late. This happens because most orders arrive well before 45 minutes, and those early deliveries pull the average down, hiding the late orders that go over the limit. Since Rosa's promise to customers is about delivering within 45 minutes, the rate of orders that break that promise tells her far more than the average does.

### II(a)
This function calculates the total cost of one late order. It adds the refund to the profit lost from future orders, which is the number of lost orders (churn) multiplied by the profit margin per order: cost = refund + churn × margin.

In [21]:
def cost_late(cost):
  refund = cost['refund']
  churn = cost['churn_orders']
  margin = cost['margin']
  total_cost = refund+margin*churn
  return total_cost
print(cost_late(COSTS))

26.2


### II(b)
This function finds the best promise for one zone and time block. For each promise in the list, it gets the delivery times, adds the profit margin for every order, and subtracts the late-order cost for every order delivered later than that promise. It keeps the promise with the highest net profit and returns that promise and its net profit.

In [22]:
def best_promise(zone, time_block, promises, costs):
  best_p = 0
  highest_profit = 0
  for promise in promises:
    total = 0
    late = 0
    orders = delivery_times(zone, time_block, promise, seed = 1)
    for order in orders:
      total += costs['margin']
      if order > promise:
        late += cost_late(costs)
    net_profit = round(total-late, 2)
    if highest_profit < net_profit:
      best_p = promise
      highest_profit = net_profit
  return best_p, highest_profit

promise = list(range(5, 80, 5))
best_p, highest_profit = best_promise("Far West", "Fri/Sat eve", promise, COSTS)
print(f"Best Promise: {best_p}min \tNet Profit: ${highest_profit}")

Best Promise: 55min 	Net Profit: $1212.4


### Choosing the range

A promise that is too short makes most orders late, and the late costs wipe out profit. A promise that is too long means fewer customers order, so profit drops. Net profit therefore rises and then falls, so the range must be wide enough to include the peak.

I tested 5 to 75 minutes in steps of 5, which covers promises clearly too short and well beyond today's 45 minutes. If the best promise landed at either edge, I would extend the range and test again.

### Recommendation

The best promise for Far West on Fri/Sat evenings is 55min, with a net profit of $1,212.40.

# AI Appendix
For Parts I and II, I used Claude (claude.ai) as a tutor, asking it to guide me without giving direct answers. It explained Python concepts, assisted me map out plain-English outlines of code structure, and pointed out bugs, which I then fixed myself. I wrote all the code in Parts I and II myself. For the markdown explanations, I drafted my answers and asked Claude to ask question if I miss something. This way I was able to develop my answers in markdowns. Then Claude helped me to check grammar and make my response clear.

For Part III, I used Claude (claude.ai) to guide me through setting up the GitHub repository, installing Claude Code, writing the skill, testing, and deploying. Claude asked me questions about which notebook functions the app needed, which cells to leave out, what inputs the app should have, and what it should show, and then drafted the `notebook-to-streamlit` SKILL.md based on my answers. Claude also suggested the prompt I used in Claude Code. Claude Code then read the skill and my notebook and generated `app.py` and `requirements.txt`.

### Key prompts
1. "I am working on assignment and I want your help guiding me in case I have any questions. You cannot give me straight answer. You are a TA and you cannot break student conduct."
2. "how do i account for all" (handling the `'all'` cases in Part I(a))
3. "How to write sorted rank"
4. "How to create promise list"
5. "Check grammar and help me with my answer to be more clear"
6. "If you think I missed something, can you ask me a question"
7. "Guide me through part 3" 
8. "Help me write skill.md" 
9. Claude Code: "Use the notebook-to-streamlit skill in .github/skills/notebook-to-streamlit/SKILL.md. Follow it to build a Streamlit app from the logic in Assignment_1.ipynb. Create app.py and requirements.txt in this folder. Don't change the notebook."
10. "I need to write AI Appendix. Give me the key prompts I used in this chat. Sammarize how I used GenAI for my assignment."